libs/isograph-react/src/core/garbageCollection.ts

```diff
@@ -51,31 +54,54 @@ export function garbageCollectEnvironment(environment: IsographEnvironment) {
     recordReachableIds(environment.store, query, retainedIds);
   }
 
-  for (const dataId in environment.store) {
-    if (!retainedIds.has(dataId)) {
-      delete environment.store[dataId];
+  for (const typeName in environment.store) {
+    const dataById = environment.store[typeName];
+    if (dataById == null) continue;
+    const retainedTypeIds = retainedIds[typeName];
+
+    // delete all objects
+    if (!retainedTypeIds) {
+      delete environment.store[typeName];
+      continue;
+    }
+
+    for (const dataId in dataById) {
+      if (!retainedTypeIds.has(dataId)) {
+        delete dataById[dataId];
+      }
+    }
+
+    if (Object.keys(dataById).length === 0) {
+      delete environment.store[typeName];
     }
   }
 }
 
+interface RetainedIds {
+  [typeName: TypeName]: Set<DataId>;
+}
+
 function recordReachableIds(
   store: IsographStore,
   retainedQuery: RetainedQuery,
-  mutableRetainedIds: Set<DataId>,
+  mutableRetainedIds: RetainedIds,
 ) {
-  recordReachableIdsFromRecord(
-    store,
-    store[ROOT_ID],
-    mutableRetainedIds,
-    retainedQuery.normalizationAst,
-    retainedQuery.variables,
-  );
+  const record =
+    store[retainedQuery.root.__typename]?.[retainedQuery.root.__link];
+  if (record)
```

Comment:
can we have `{` here? And in general, we should have them everywhere, so as to be clear and idiomatic.

Should we throw if the record is not present? Not quite! The record is allowed to `=== null`, but it cannot `=== undefined` (since it's retained.)

Reply:
In both of these cases it would make sense to warn in dev about bug in Isograph with garbage collection